### DISCUSSION TITLE
Best practice: should Symfony services be stateless?

### DISCUSSION THREAD

**Discussion by igoooor** [symfony/symfony#42858] [Q&A] [answered]
It's all in the title.
I always had in my mind that services should be stateless. I'm sure that came from learning symfony with more experienced developers back then, but I actually never challenged that idea.
Is it something correct? or complete BS?

**Comment by stof** [comment, marked as answer]
yes, it is indeed a best practice. In most cases, making the service stateful creates either weird bugs (what if a service gets called recursively ? Making a stateful service re-entrant is often harder than making it stateless) or memory leaks.

Of course, some services need to be stateful due to their function (for instance, it would not make sense for the Doctrine EntityManager to be stateless, as it is responsible for managing the state of entities)
